### ISSUE TITLE
linked properties: \prop_get:cnNTF errors if property is unknown

### ISSUE THREAD

**Issue by u-fischer** [latex3/latex3#1455]
This document works fine upto now (without the new linked properties) and printed a b:

~~~~
\documentclass{article}
\begin{document}
xxx
\ExplSyntaxOn
\prop_get:cnNTF {g_some_unknown_prop} {blub}\l_tmpa_tl{a}{b}
\ExplSyntaxOff
\end{document} 
~~~~

With the new code it now errors

~~~~
! Use of \__prop_get_linked:w doesn't match its definition.
<recently read> \g_some_unknown_prop 
                                     
l.192 ...TF {g_some_unknown_prop} {blub}\l_tmpa_tl
                                                  {a}{b}
~~~~

Similar `\prop_gput:cnn {g_some_unknown_prop}{a}{b}` that up-to-now was silently swallowed but now errors with the same error.

**Comment by josephwright** [comment]
Checking the docs, we never say you can use `\prop_get:NnN` with undefined `prop` vars - we only say what happens if the key does not exist in the `prop`. I don't think the old behaviour was by-design, more an implementation detail.

**Comment by josephwright** [comment]
Of course, we should not give the error shown: at worse, we should give an undefined csname for the variable.Somewhat related is https://github.com/latex3/latex3/issues/1458, as that is also about changes due to the need to have `prop` defined (though that one really isn't somewhere we need a change I suspect).

**Comment by blefloch** [comment]
I see that the previous code was very permissive.  What transitional behavior do we want here?  Silent False branch, or reasonable error plus false branch?

The consensus so far had been that any wrong user code can give arbitrary errors and that we should mostly optimize for speed instead.  But it seems from the various bug reports that this is not what we want?

**Comment by blefloch** [comment]
That said, `\ExplSyntaxOn\int_set:cn{l_unknown_int}{1}` gives a pretty gnarly set of errors,
```
! LaTeX Error: Missing \begin{document}.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
<*> \ExplSyntaxOn\int_set:cn{l_unknown_int}{1}
                                              ^^M
? 
Missing character: There is no = in font nullfont!
! You can't use `\numexpr' in horizontal mode.
\int_set:Nn #1#2->#1=\__int_eval:w 
                                   #2\__int_eval_end: 
<*> \ExplSyntaxOn\int_set:cn{l_unknown_int}{1}
                                              ^^M
? 
Missing character: There is no 1 in font nullfont!
```

**Comment by josephwright** [comment]
> The consensus so far had been that any wrong user code can give arbitrary errors and that we should mostly optimize for speed instead. But it seems from the various bug reports that this is not what we want?

I think on balance we do want speed, but when we make a change we have to be sure that it doesn't result in cases that passed before now failing.

**Comment by blefloch** [comment]
So here you would say we should have no error at all for `\prop_get` or `\prop_put` on a prop that is `\relax`?  Bug-for-bug compatibility is going to be pretty annoying to reach (but doable I suppose).

**Comment by josephwright** [comment]
@blefloch  More that we should have known we were changing that, and warned about it - I think we need to extend test coverage, mainly

**Comment by muzimuzhi** [comment]
Is "`\prop_get` or `\prop_put` on a prop that is `\relax`" defined behavior?

**Comment by josephwright** [comment]
@blefloch  I'm going to first-aid any packages that don't get updated by the end of this week

**Comment by muzimuzhi** [comment]
> No, it's undefined behaviour, ... .

Then I doubt test should be extended to cover even undefined behaviors.

From a user aspect (a user of `l3kernel`), getting a detailed error message only once for each different cases of undefined behavior might be a balance between speed and usefulness, but that complicates implementation.

**Comment by Skillmon** [comment]
I'd say people exploiting implementation details deserve no sympathy (which includes me... I should really run my test suites). With `l3debug` it's another matter and one can implement all the checks we can think of, but outside of debugging, speed should be more important than catching errors due to changed implementation details.
